sdstore/sdstore_test.go

```diff
@@ -602,3 +633,87 @@ func TestZipAndUnzipWithSymlink(t *testing.T) {
 	os.RemoveAll("../data/test")
 	os.RemoveAll("../data/testsymlink.zip")
 }
+
+func TestBackOff(t *testing.T) {
+	client := newStore(1.5, 5)
+	cases := []struct {
+		attemptNum  int
+		backoffTime time.Duration
+	}{
+		{attemptNum: 1, backoffTime: 1 * time.Second},
+		{attemptNum: 5, backoffTime: 37 * time.Second},
+	}
+
+	for _, c := range cases {
+		bt := client.backOff(c.attemptNum)
+		assert.Equal(t, c.backoffTime, bt, fmt.Sprintf("when attempt number is %v", c.attemptNum))
+		if bt != c.backoffTime {
+			t.Errorf("Expected time is %q. But Got %q", c.backoffTime, bt)
+		}
+	}
+}
+
+func TestCheckForRetry(t *testing.T) {
+	client := newStore(1.0, 3)
+
+	cases := []struct {
+		err        error
+		statusCode int
+		doesRetry  bool
+		retryErr   error
+	}{
+		{err: fmt.Errorf("awful things happen"), statusCode: 0, doesRetry: true, retryErr: fmt.Errorf("awful things happen")},
+		// statut 200~
+		{statusCode: http.StatusOK, doesRetry: false},
+		{statusCode: http.StatusCreated, doesRetry: false},
+		{statusCode: http.StatusAccepted, doesRetry: false},
+		{statusCode: http.StatusNoContent, doesRetry: false},
+		// status 400~
+		{statusCode: http.StatusBadRequest, doesRetry: true},
+		{statusCode: http.StatusUnauthorized, doesRetry: true},
+		{statusCode: http.StatusForbidden, doesRetry: true},
+		{statusCode: http.StatusNotFound, doesRetry: false, retryErr: fmt.Errorf("got 404 Not Found. stop retring")},
+		{statusCode: http.StatusRequestTimeout, doesRetry: true},
+		// status 500~
+		{err: nil, statusCode: http.StatusInternalServerError, doesRetry: true},
+	}
+
+	for _, c := range cases {
+		res := &http.Response{}
+		res.StatusCode = c.statusCode
+		res.Status = "404 Not Found"
+		retry, err := client.checkForRetry(res, c.err)
+		assert.Equal(t, c.doesRetry, retry, fmt.Sprintf("when status is %d and err is %q", c.statusCode, c.err))
+		assert.Equal(t, c.retryErr, err, "when status is %d and err is %q", c.statusCode, c.err)
+	}
+}
+
+func TestDo(t *testing.T) {
```

Comment:
I think it's better to change this test to a table driven test.

Reply:
I try to do table driven test.

Reply:
Done!